yarn-project/noir-compiler/src/contract-interface-gen/noir.ts

```diff
@@ -0,0 +1,247 @@
+ */
+function generateStaticImports() {
+  return `use dep::std;
+use dep::aztec::context::Context;
+use dep::aztec::constants_gen::RETURN_VALUES_LENGTH;`;
+}
+
+/**
+ * Generate the main focus of this code generator: the contract interface struct.
+ * @param contractName - the name of the contract, as matches the original source file.
+ * @returns Code.
+ */
+function generateContractInterfaceStruct(contractName: string) {
+  return `struct ${contractName}ContractInterface {
+  address: Field,
+}
+`;
+}
+
+/**
+ * Generates the implementation of the contract interface struct.
+ * @param contractName - The name of the contract, as matches the original source file.
+ * @param functions - An array of strings, where each string is valid Noir code describing the function interface of one of the contract's functions (as generated via `generateFunctionInterface` above).
+ * @returns Code.
+ */
+function generateContractInterfaceImpl(contractName: string, functions: string[]) {
+  return `impl ${contractName}ContractInterface {
+  fn at(address: Field) -> Self {
+      Self {
+          address,
+      }
+  }
+  ${functions.join('\n')}
+}
+`;
+}
+
+/** Represents a struct along its parent names to derive a fully qualified name. */
+type StructInfo = ABIVariable & { /** Parent name */ parentNames: string[] };
+
+/**
+ * Generates a Noir struct.
+ * @param struct - Struct info.
+ * @returns Code representing the struct.
+ */
+function generateStruct(struct: StructInfo) {
+  const fields = (struct.type as StructType).fields.map(
+    field => `  ${field.name}: ${getTypeName(field, struct.name, ...struct.parentNames)},`,
+  );
+
+  return `
+struct ${getStructName(struct.name, ...struct.parentNames)} {
+${fields.join('\n')}
+}`;
+}
+
+/**
+ * Collects all structs across all parameters.
+ * @param params - Parameters to look for structs, either structs themselves or nested.
+ * @param parentNames - Parent names to derive fully qualified names when needed.
+ * @returns A list of struct infos.
+ */
+function collectStructs(params: ABIVariable[], parentNames: string[]): StructInfo[] {
+  const structs: StructInfo[] = [];
+  for (const param of params) {
+    if (param.type.kind === 'struct') {
+      const struct = { ...param, parentNames };
+      structs.push(struct, ...collectStructs(param.type.fields, [param.name, ...parentNames]));
+    } else if (param.type.kind === 'array') {
+      structs.push(...collectStructs([{ name: param.name, type: param.type.type }], [...parentNames]));
+    }
+  }
+  return structs;
+}
+
+/**
+ * Generates the Noir code to represent an interface for calling a contract.
+ * @param abi - The compiled Noir artifact.
+ * @returns The corresponding ts code.
+ */
+export function generateNoirContractInterface(abi: ContractAbi) {
+  const contractStruct: string = generateContractInterfaceStruct(abi.name);
+  const methods = compact(abi.functions.filter(f => f.name !== 'constructor' && !f.isInternal));
```

Comment:
We might want to include internal functions in the interface. Iirc, `internal` here is used to label public internal functions. A public internal function -- when called by a private function -- is not inlined: it's executed as a separate kernel iteration (because the sequencer must execute it). So we probably _do_ want to be able to make calls to public internal functions via a pretty interface. `my_public_internal_function.foo(...)`.

Reply:
As I started coding this, I ended up rolling it back, since it generates a very annoying cyclic dependency: you need contract A to compile in order to generate its interface which you need to use from within contract A to call itself, but contract A doesn't compile unless you have its interface since you're using it as part of its code.

We need Noir macros.

Reply:
Oh dear! 
So, for now at least, a call to an internal public function will need to be done via the less pretty syntax of calling a method of the context, and passing a manually-computed selector? Sad face.

One thing we could do (to avoid the need to manually generate a selector), is write an noir unconstrained function which derives the function selector from a function signature. We’d need to request a string_to_bytes function from the Noir team. With this, it would be more human readable, at least.

> You need contract A to compile in order to generate its interface

I guess that’s not strictly true (we could parse the code separately from the compiler), but I agree it’s true given that we’ve taken the approach of relying on the json abi to derive the contract interface.

How would macros help, in this case?

Reply:
> How would macros help, in this case?

Just the way you say! Since a macro would work with the AST directly, it could generate the interface function without requiring the whole contract to be compiled. The macro could even be used for generating the function _within_ the same contract if it's internal.

Reply:
I opened an issue here, to continue the discussion: https://github.com/AztecProtocol/aztec-packages/issues/1550